clio-client/src/main/scala/org/broadinstitute/clio/client/util/IoUtil.scala

```diff
@@ -47,34 +50,36 @@ trait IoUtil {
    * google-cloud-nio adapter will eagerly check storage authentication
    * for each path, and fail.
    */
-  import IoUtil.GsUtil
 
   def readGoogleObjectData(location: URI): String = {
-    val gs = new GsUtil(None)
-    gs.cat(location.toString)
+    gsUtil.cat(location.toString)
   }
 
   def copyGoogleObject(from: URI, to: URI): Int = {
-    val gs = new GsUtil(None)
-    gs.cp(from.toString, to.toString)
+    gsUtil.cp(from.toString, to.toString)
   }
 
   def deleteGoogleObject(path: URI): Int = {
-    val gs = new GsUtil(None)
-    gs.rm(path.toString)
+    gsUtil.rm(path.toString)
   }
 
   def googleObjectExists(path: URI): Boolean = {
-    val gs = new GsUtil(None)
-    gs.exists(path.toString) == 0
+    gsUtil.exists(path.toString) == 0
   }
 
-  def getMd5HashOfGoogleObject(path: URI): Symbol = {
-    Symbol(new GsUtil(None).md5Hash(path.toString))
+  private val md5HashPattern = "Hash \\(md5\\):\\s+([0-9a-f]+)".r
+
+  def getMd5HashOfGoogleObject(path: URI): Option[Symbol] = {
+    /*
+     * Files uploaded through parallel composite uploads won't have an md5 hash.
```

Comment:
What is the expected behavior if google doesn't have an MD5 hash for a file? Do we need to manually generate a hash in that case? Will the database/file MD5 be used instead?

How many files are affected by this? Google docs say any file uploaded in parallel will be affected, it seems like that would apply to all CRAM/BAM etc files.

Reply:
If google doesn't have an md5, and there's no md5 in the database during backfill, then Clio won't know the md5 for the file. As far as I've seen, this is going to affect all ubams (because we use the parallel uploads in push-to-cloud), but not gvcfs or crams (maybe because they're generated in-cloud?).

We were willing to live without ubam md5s when we released Clio phase 1, so the plan is to just treat the md5 as optional in the backfill. Ditto for gvcfs, since their md5s aren't delivered. For crams, the backfill will check that we got an md5 from somewhere, and raise an error if not.